tests/_support/AcceptanceTester.php

```diff
@@ -18,9 +18,45 @@
 */
 class AcceptanceTester extends \Codeception\Actor
 {
-    use _generated\AcceptanceTesterActions;
+	use _generated\AcceptanceTesterActions;
 
-    /**
-     * Define custom actions here
-     */
+	/**
+	 * Custom actions here
+	 */
+
+	/**
+	 * Trigger actions during remote acceptance test executions.
+	 *
+	 * @param string $file File path.
+	 * @param callable $function Reference to a static class method.
+	 * @param bool $rollback_on_shutdown Rollback changes on shutdown.
+	 *
+	 * @throws Exception Throws an exception if passed parameters are not as expected.
+	 *
+	 * @return void
+	 */
+	public function preLoadConfig( string $file, callable $function, bool $rollback_on_shutdown = true ) {
+		$path = __DIR__ . '/../../../../tests/load.php';
+
+		if ( ! file_exists( $file ) ) {
+			throw new Exception( 'File path does not exist.' );
+		}
+
+		if ( ! is_array( $function ) || ! is_string( $function[0] ) ) {
+			throw new Exception( 'Passed function needs to be a static method of a class.' );
```

Comment:
Why have this limitation?

Reply:
So the method can be called statically without bootstrapping or instantiating the class. Helps remind the developer that they don't have full access to the test thread that schedules the execution on the acceptance request/thread.

Reply:
I couldn't use a lambda function or regular function this way which would also work.

Reply:
This also assumes the callable is in the same class as the tests but there's nothing to enforce that so it could be some outside instantiated class. I could also reference a static method like `SomeClass::method` as a string.

I'd personally either loosen the constraint or see if it's possible to further force that intentionality, e.g. it has to be a static method of the calling class.

Reply:
So, had time to think about this further and I have no idea how a lambda would behave when converted to a string! 

That said I'm wondering if we can or should remove the need to pass the `__FILE__` and `__CLASS__` parameters. In this implementation those are entirely connected so every time the method is used it'd have to be the same parameters. Could we derive those in this method? Then to use the `preLoadConfig()` method the args would just be static method name and rollback...

Reply:
😬 `debug_backtrace()` might be the only option.

Reply:
- We can't use a lambda since it is only defined within the context of the running thread, not the thread where the acceptance test will be running in. Similar restriction to why we're discouraging use of non-static methods.
- I only added the file and class params so we're not strictly limited to same-class methods, however, now that you mentioned it, I think there is not much effort in creating a method in the same class which then runs any other type of callable from there.
- And yes, we can either use reflections or debug_backtrace to know the class/file of the calling method, will see if methods being in a trait would provide any advantage in that regard.

Reply:
Seems like __CLASS__ would reference the actual test class not the trait, combining that with reflections, we can get the file name. So that could work. eg:

```php
$class_name = __CLASS__;
$reflector = new \ReflectionClass( $class_name );
$file_name = $reflector->getFileName();
```

However, the only caveat there would be if someone extended that class, quite distant change that might happen though. So I believe we can go through with it. Thoughts ?

Reply:
Also, what do you think about changing the method name to `haveCode`, so it is written like `$I->haveCode( $method )`, or anything of similar meaning.. ? 🤔

Reply:
The `$I` specific method naming makes sense. Maybe `$I->haveConfig()` or `$I->bootstrapWith()`? Lots of things it could be!

> However, the only caveat there would be if someone extended that class, quite distant change that might happen though. So I believe we can go through with it. Thoughts ?

Yeah requiring the file in that way could break easily unless reflection can be used to require the full chain. Still wondering if there might be other approaches...

`$I->haveHttpHeader()` and `$I->setServerParameters([])` might be of some use.

Reply:
`$I->bootstrapWith()` looks nice!

I don't think `setServerParameters` would work with web drivers, since CC is not even loaded there. It is specific to PhpBrowser if I'm not mistaken.

I'll try to improve the file name detection through reflections, and provide an optional param to override it if needed for those edge cases where inheritance is in play.

Reply:
Ah, also re the trait thing I mentioned, ignore that, since the Actor is not really a trait, it's a whole different object than the test class. So what we have now is the best I could think of, using reflections, since the actor doesn't really have access to the test class/method itself.

Reply:
Aha, didn't see I was looking at the PHPBrowser versus WebDriver methods 🤦‍♂️ 

Although I really like having the bootstrap code alongside the tests in the class I think my hesitance is that idiomatically, like in the case of regular PHPUnit tests, this sort of thing is done using a bootstrap file.

If there is inheritance in play and you need to specify a file, all that file will be doing is requiring the necessary files in order for the static method to be called successfully. It might be tricky to figure out which files to require. At that point you have both the static method, and a bootstrap file.

You could instead just require a file e.g. `$I->bootstrapWith( 'tests/acceptance/bootstrap.php' );`.

Some advantages of that approach:
- full flexibility of an arbitrary file
- no concerns about necessary files not being included, can just define constants etc directly
- simpler interface

Downsides:
- not having the bootstrapping code in the same file alongside tests

What do you think?

Would appreciate you thoughts too @rmccue, don't want to hold this up any longer.

Reply:
I'm unsure using files instead of methods has a lot advantages, I think the same-class method approach is simpler, and you can add whatever code you need to the method as you would with the arbitrary file, also it's a bit more organized vs a file for each test that needs it.

This isn't necessarily _bootstrapping_ code, it'd mostly be _overriding_ code instead. The aim here is to be able to add actions and filters to modify the default behavior to be able to prep the environment for a specific test.

Reply:
@shadyvb Ryan found this package called that allows you to serialize closures: https://opis.io/closure/3.x/serialize.html

I think that would give us the nicest interface so you could straight up use `$I->bootstrapWith( function () { ... } );`. Sound ok?

Reply:
Sorry, feels like going round in circles a bit here, don't worry about the closure thing.

Let's go with this and see how it goes in use. I'll add those commit suggestions.